neo4j/auth/cert.go

```diff
@@ -0,0 +1,162 @@
+}
+
+// NewRotatingClientCertificateProvider creates a new RotatingClientCertificateProvider given a ClientCertificate.
+// This function loads the certificate-key pair specified in the ClientCertificate and returns
+// a provider that allows updating the certificate dynamically through UpdateCertificate method.
+//
+// NewRotatingClientCertificateProvider is part of the mTLS preview feature
+// (see README on what it means in terms of support and compatibility guarantees)
+func NewRotatingClientCertificateProvider(cert ClientCertificate) (*RotatingClientCertificateProvider, error) {
+	tlsCert, err := loadCertificate(cert.CertFile, cert.KeyFile, cert.Password)
+	if err != nil {
+		return nil, err
+	}
+	return &RotatingClientCertificateProvider{certificate: &tlsCert}, nil
+}
+
+func (p *RotatingClientCertificateProvider) GetCertificate() *tls.Certificate {
+	p.mu.RLock()
+	defer p.mu.RUnlock()
+	return p.certificate
+}
+
+// UpdateCertificate updates the certificate stored in the provider with a new certificate specified
+// by the ClientCertificate. This method allows dynamic updates to the certificate used in TLS connections,
+// facilitating use cases such as certificate rotation.
+//
+// UpdateCertificate is part of the mTLS preview feature
+// (see README on what it means in terms of support and compatibility guarantees)
+func (p *RotatingClientCertificateProvider) UpdateCertificate(cert ClientCertificate) error {
+	tlsCert, err := loadCertificate(cert.CertFile, cert.KeyFile, cert.Password)
+	if err != nil {
+		return err
+	}
+
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	p.certificate = &tlsCert
+	return nil
+}
+
+// loadCertificate attempts to load a certificate and its corresponding private key
+// from the given paths. If a password is provided, it will attempt to decrypt
+// the private key assuming it is encrypted.
+func loadCertificate(certFile string, keyFile string, password *string) (tls.Certificate, error) {
+	// Load certificate PEM block
+	certPEMBlock, err := os.ReadFile(certFile)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+	// Load private key PEM block
+	keyPEMBlock, err := os.ReadFile(keyFile)
+	if err != nil {
+		return tls.Certificate{}, err
+	}
+	// Decrypt the private key if a password is provided
+	if password != nil {
+		decodedKeyPEMBlock, _ := pem.Decode(keyPEMBlock)
+		if decodedKeyPEMBlock == nil {
+			return tls.Certificate{}, errors.New("failed to parse PEM block containing the key")
+		}
+		//lint:ignore SA1019 Using due to lack of stdlib alternatives; aware of RFC 1423's insecurity.
+		decryptedDERBlock, decryptErr := x509.DecryptPEMBlock(decodedKeyPEMBlock, []byte(*password))
+		if decryptErr != nil {
+			return tls.Certificate{}, decryptErr
+		}
+		// Re-encode the decrypted block to PEM
+		keyPEMBlock = pem.EncodeToMemory(&pem.Block{Type: decodedKeyPEMBlock.Type, Bytes: decryptedDERBlock})
+	}
+	// Create the tls.Certificate
+	return tls.X509KeyPair(certPEMBlock, keyPEMBlock)
```

Comment:
I don't understand this line of code. By only returning one value here instead of a tuple, do you get `(theReturnValue, nil)`?

Reply:
Hey! The signature for the function is `func X509KeyPair(certPEMBlock, keyPEMBlock []byte) (Certificate, error)`, it returns and error (if any) with the certificate 👍

Reply:
Ahhh, I see. Thanks!